consensus/src/lib.rs

```diff
@@ -114,74 +113,79 @@ impl<'a> ConsensusRef<'a> {
 
     fn reorganize(
         &mut self,
-        cached_inputs: &mut CachedInputs,
         best_block_id: Option<Id<Block>>,
         top_block_index: &BlockIndex,
     ) -> Result<(), BlockError> {
-        // Disconnect the current chain
+        // Disconnect the current chain if it is not a genesis
         if let Some(ref best_block_id) = best_block_id {
             let mut ancestor = self
                 .db_tx
                 .get_block_index(&best_block_id)?
                 .expect("Can't get block index. Inconsistent DB");
             while !self.is_block_in_main_chain(&ancestor) {
-                self.disconnect_tip(cached_inputs, &mut ancestor)?;
+                self.disconnect_tip(&mut ancestor)?;
                 ancestor = self.get_ancestor(ancestor.get_block_id())?;
             }
         }
 
         // Connect the new chain
         let mut ancestor = top_block_index.clone();
         while !self.is_block_in_main_chain(&ancestor) {
-            self.connect_tip(cached_inputs, &ancestor, &best_block_id)?;
+            self.connect_tip(&ancestor, &best_block_id)?;
             ancestor = self.get_ancestor(ancestor.get_block_id())?;
         }
         Ok(())
     }
 
-    fn store_cached_inputs(&mut self, cached_inputs: &mut CachedInputs) -> Result<(), BlockError> {
+    fn store_cached_inputs(&mut self, cached_inputs: &CachedInputs) -> Result<(), BlockError> {
         for (tx_id, tx_index) in cached_inputs.iter() {
             self.db_tx.set_mainchain_tx_index(tx_id, tx_index)?;
         }
         Ok(())
     }
 
-    fn connect_transactions(
-        &mut self,
-        cached_inputs: &mut CachedInputs,
-        transactions: &Vec<Transaction>,
-        best_block_id: &Option<Id<Block>>,
-    ) -> Result<(), BlockError> {
-        let mut byte_offset_in_block = 0;
-        let mut serialized_size = 0;
-        for tx in transactions.iter() {
-            serialized_size += std::mem::size_of_val(tx) as u32;
+    fn calculate_indices(
+        &self,
+        block: &Block,
+        tx: &Transaction,
+    ) -> Result<TxMainChainIndex, BlockError> {
+        let enc_block = block.encode();
+        let enc_tx = tx.encode();
+        let offset_tx = enc_block
+            .windows(enc_tx.len())
+            .enumerate()
+            .find_map(|(i, d)| (d == enc_tx).then(|| i))
+            .ok_or(BlockError::Unknown)? as u32;
 
-            dbg!(tx);
-            let inputs = tx.get_inputs();
-            for input in inputs {
-                let input_index = input.get_outpoint().get_output_index();
+        let tx_position =
+            TxMainChainPosition::new(&block.get_id().get(), offset_tx as u32, enc_tx.len() as u32);
```

Comment:
We already cast `offset_tx` to u32 in its declaration, no need to cast it again.

Reply:
Correct. And I'll even add you one. I don't like casting. A try_into() is much better.

Reply:
Fixed!